docs/RFCS/20222202_historical_hot_ranges.md

```diff
@@ -0,0 +1,460 @@
+store, query, and render. To reduce the number of ranges we collect metrics for,
+we can set a budget of 1000 max ranges. To populate these 1000 ranges we can use
+a mixed strategy of filtering out the coldest ranges and aggregating what
+remains. This is discussed in detail later on.
+
+To provide meaningful resolution, we can set a sampling rate of 4 times per
+hour, and store samples for a period of 14 days.
+
+For each range, here is the information captured: 
+| Value      | Bytes   |
+| ---------- | ------- |
+| range id   | 4       |
+| replica id | 4       |
+| qps        | 4       |
+| start key  | <= 100  |
+| end key    | <= 100  |
+| node ids   | ~12     |
+| store ids  | ~12     |
+| schema     | ~20     |
+| locality   | ~20     |
+| rangeBytes | 4       |
+
+The bytes required to store node ids, store ids, schema, and locality are
+estimates, as these values can vary in practice. The bytes required to start and
+end keys could be much less depending on the distribution of start and end key
+character lengths.
+
+
+We can calculate the total storage requirements of this data, less the cost of
+the indexes and a unique identifier, considering all dimensions: ``` ~ 270 bytes
+per range * 1000 ranges per sample * 4 samples per hour * 24 hours * 14 days =
+~363 Megabytes ```
+
+
+It's important to point out that the client will never need the majority of
+these bytes to be sent over the wire in order to render the visualization.
+Instead, the client will request samples in chunks, with subsequent API calls
+made to request the specific details of a range only as needed. This is
+discussed in more detail later on.
+
+### Metric Generation 
+
+The metric generation scheme is designed for both single and multi-tenant
+deployments, and operates with two different sampling modes: 
+#### Tenant Sampling 
+Tenant sampling generates metrics for a specific tenant. Tenant
+sampling is performed by the host, and metrics are stored such that a tenant can
+only access metrics for ranges inside its own keyspace. This prevents a tenant from
+accessing hotness from system-level ranges or another tenant's ranges. Every
+tenant receives its own 1000 range budget to prevent colder tenants from being
+overshadowed by hotter tenants.
+
+#### Host Sampling 
+Host sampling generates metrics reported over the cluster's
+entire keyspace in a tenant-agnostic way. The host receives its own 1000 range
+budget.
+
+A dedicated single-tenant deployment only performs host sampling, while
+multi-tenant deployments perform both tenant and host sampling. In a
+single-tenant deployment, the results of tenant's sampling sampling would be
+contained in the host sampling, and therefore be redundant.
+
+Tenant sampling and Host sampling follow mostly the same process - the
+differences manifest in how aggregation is performed to keep samples within
+budget, and how the sampled metrics are written to `system.hot_ranges`.
+
+According to the sample rate, a job is executed to
```

Comment:
At what frequency is the job collecting data? Can we make this configurable?

Reply:
The job runs every 15 minutes. I don't know if it would be in the first iteration, but the sample frequency could be a cluster setting.

Reply:
Is the plan to use scheduled jobs? Do you think that'd be better than just perpetually running a job which sits and sleeps?

Reply:
> Is the plan to use scheduled jobs?

Yes

> Do you think that'd be better than just perpetually running a job which sits and sleeps?

When you say 'job' are you referring to `pkg/job`?  I want to make sure that if that if a node dies, the sampling isn't cancelled or interrupted, and my understanding that `pkg/job` make sure of all that.

Reply:
Yes. For better or worse, both the regular "job" and "scheduled jobs" are in the same package. The scheduled job framework launches a job periodically. I'd recommend you consider using just a singleton job rather than a scheduled job. I think it'll be simpler and more flexible.

Reply:
Ok we'll check it out!